src/modules/cpu/kernel/rain.hpp

```diff
@@ -0,0 +1,1203 @@
+#include "rpp_cpu_common.hpp"
+
+// Constants to represent the rain intensity for different data types
+#define RAIN_INTENSITY_8U 200   // Intensity value for Rpp8u
+#define RAIN_INTENSITY_8S 72    // Intensity value for Rpp8s
+#define RAIN_INTENSITY_FLOAT 200 * ONE_OVER_255 // Intensity value for Rpp32f and Rpp16f
+
+template<typename T>
+inline void create_rain_layer(T *rainLayer, Rpp32f rainPercentage, RpptDescPtr srcDescPtr, Rpp32s slant, Rpp32u dropLength, Rpp32u rainWidth)
+{
+    Rpp32f rainPercent = rainPercentage * 0.004f;
+    Rpp32u numDrops = static_cast<Rpp32u>(rainPercent * srcDescPtr->h * srcDescPtr->w);
+
+    // Seed the random number generator and set up the uniform distributions
+    std::mt19937 rng(std::random_device{}());
+    std::uniform_int_distribution<Rpp32u> distX(0, srcDescPtr->w - slant - 1);
+    std::uniform_int_distribution<Rpp32u> distY(0, srcDescPtr->h - dropLength - 1);
+
+    // Choose the rain intensity value based on the data type
+    T rainValue = std::is_same<T, Rpp8u>::value ? static_cast<T>(RAIN_INTENSITY_8U) :
+                  std::is_same<T, Rpp8s>::value ? static_cast<T>(RAIN_INTENSITY_8S) :
+                  static_cast<T>(RAIN_INTENSITY_FLOAT);
+    Rpp32f slantPerDropLength = static_cast<Rpp32f>(slant) / dropLength;
+    for (Rpp32u i = 0; i < numDrops; i++)
+    {
+        Rpp32u xStart = distX(rng);
+        Rpp32u yStart = distX(rng);
+        for (Rpp32u j = 0; j < dropLength; j++)
+        {
+            Rpp32u x = xStart + j * slantPerDropLength;
+            Rpp32u y = yStart + j;
+
+            if ((x >= 0) && (x < srcDescPtr->w) && (y < srcDescPtr->h))
+            {
+                T *rainLayerTemp = rainLayer + y * srcDescPtr->w + x;
+                for (Rpp32u k = 0; k < rainWidth; k++)
+                    rainLayerTemp[k] = rainValue;
+            }
+        }
+    }
+}
+
+RppStatus rain_u8_u8_host_tensor(Rpp8u *srcPtr,
+                                 RpptDescPtr srcDescPtr,
+                                 Rpp8u *dstPtr,
+                                 RpptDescPtr dstDescPtr,
+                                 Rpp32f rainPercentage,
+                                 Rpp32u rainWidth,
+                                 Rpp32u rainHeight,
+                                 Rpp32s slant,
+                                 Rpp32f *alphaValues,
+                                 RpptROIPtr roiTensorPtrSrc,
+                                 RpptRoiType roiType,
+                                 RppLayoutParams layoutParams,
+                                 rpp::Handle& handle)
+{
+    RpptROI roiDefault = {0, 0, (Rpp32s)srcDescPtr->w, (Rpp32s)srcDescPtr->h};
+    Rpp32u numThreads = handle.GetNumThreads();
+
+    Rpp8u *rainLayer = reinterpret_cast<Rpp8u *>(handle.GetInitHandle()->mem.mcpu.scratchBufferHost);
+    std::memset(rainLayer, 0, srcDescPtr->w * srcDescPtr->h * sizeof(Rpp8u));
+    create_rain_layer(rainLayer, rainPercentage, srcDescPtr, slant, rainHeight, rainWidth);
```

Comment:
What percentage of time do just these 3 lines L86-L88 take as compared to the whole function?

Reply:
around 50-60% on average